File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        List<TreeNode> path_p = pathToNode(root, p);
        List<TreeNode> path_q = pathToNode(root, q);
        TreeNode ancestor = root;
        for (int i = 0; i < Math.min(path_p.size(), path_q.size()); i++) {
            if (path_p.get(i).val != path_q.get(i).val) {
                return ancestor;
            }
            ancestor = path_p.get(i);
        }
        return ancestor;
    }

    private List<TreeNode> pathToNode(TreeNode root, TreeNode n) {
        TreeNode curr = root;
        List<TreeNode> path = new ArrayList<>();
        path.add(curr);
        while (curr.val != n.val){
            if (n.val > curr.val){
                curr = curr.right;
            }
            else {
                curr = curr.left;
            }
            path.add(curr);
        }

        return path;
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        """
        Plan of action: we could do an in order traversal and mark the nodes. Just running DFS should work.
        """
        curr_index = 0
        val = -1
        
        def dfs(node: TreeNode) -> None:
            nonlocal curr_index, val
            
            if not node:
                return

            dfs(node.left)
            curr_index+=1
            if curr_index == k:
                val = node.val
                return
            dfs(node.right)            

        dfs(root)
        return val